File: python/day2.py

def is_decreasing_safe(line):
    return all((x < y) and (abs(x-y) <= 3) for x, y in zip(line[:-1], line[1:]))

def is_increasing_safe(line):
    return all((x > y) and (abs(x-y) <= 3) for x, y in zip(line[:-1], line[1:]))

def part1(lines):
    num_safe = 0
    for line in lines:
        num_safe += (is_decreasing_safe(line) or is_increasing_safe(line))
    print(num_safe)

def part2(lines):
    num_safe = 0
    for line in lines:
        is_decreasing_safe_remove_1 = any(
            is_decreasing_safe(line[:i] + line[i+1:]) for i in range(len(line))
        ) or is_decreasing_safe(line)
        is_increasing_safe_remove_1 = any(
            is_increasing_safe(line[:i] + line[i+1:]) for i in range(len(line)) 
        ) or is_increasing_safe(line)

        num_safe += (is_decreasing_safe_remove_1 or is_increasing_safe_remove_1)
    print(num_safe)

if __name__ == '__main__':

    with open("inputs/day2.txt") as f:
        lines = f.readlines()
        lines = [[int(x) for x in line.split(" ")] for line in lines]
    
    part1(lines)
    part2(lines)

File: scala/day2.scala
//> using scala 3.5.2
//> using toolkit default

package day2
def readInput() =
    val path: os.Path = os.pwd / "inputs" / "day2.txt"
    val lines: Seq[String] = os.read.lines(path)
    val linesFormatted: Seq[Seq[Int]] = lines.map(
        _.split(" ").map(_.toInt)
    )
    linesFormatted

def isDecreasingAndSafe(line: Seq[Int]): Boolean =
    val result = (line.take(line.length-1) zip line.takeRight(line.length-1)).map{
        case (x, y) => (x < y && ((x - y).abs <= 3))
    }.foldLeft(true)(_ && _)
    result

def isIncreasingAndSafe(line: Seq[Int]): Boolean =
    val result = (line.take(line.length-1) zip line.takeRight(line.length-1)).map{
        case (x, y) => (x > y && ((x - y).abs <= 3))
    }.foldLeft(true)(_ && _)
    result

def isDecreasingAndSafeRemoveOne(line: Seq[Int]) =
    if isDecreasingAndSafe(line) then
        true
    else
        (0 to line.length).map(
            i => line.take(i) ++ line.takeRight(line.length - i - 1)
        ).map(isDecreasingAndSafe).foldLeft(false)(_ || _)

def isIncreasingAndSafeRemoveOne(line: Seq[Int]) =
    if isIncreasingAndSafe(line) then
        true
    else
        (0 to line.length).map(
            i => line.take(i) ++ line.takeRight(line.length - i - 1)
        ).map(isIncreasingAndSafe).foldLeft(false)(_ || _)


def part1(lines: Seq[Seq[Int]]) =
    var numSafe = 0
    for line <- lines do
        if (isIncreasingAndSafe(line) || isDecreasingAndSafe(line)) then
            numSafe += 1
    println(numSafe)

def part2(lines: Seq[Seq[Int]]) = 
    var numSafe = 0
    for line <- lines do
        if (isIncreasingAndSafeRemoveOne(line) || isDecreasingAndSafeRemoveOne(line)) then
            numSafe += 1
    println(numSafe)

@main def day2() =
    val lines: Seq[Seq[Int]] = readInput()
    part1(lines)
    part2(lines)
